Payments contract first draft

aarshkshah1992 opened the pull request:
This is the first draft for the payments contract MVP. It has the following features: 

- ERC-20 deposits
- ERC-20 withdrawals
- rail creation
- operator approvals
- operator termination
- rail payment modification
- rail lockup modification
- lazy rail settlement
- proxy pattern for contract upgrade
- arbitration
- rate change queuing
- avoiding debt creation during lockup or rate modification


All rail modifications and creation respect the operator approval.

The following features which will be added in a subsequent PR:

1) Escrow
2) Payment Fees

Stebalien commented on README.md:
```diff
@@ -1,3 +1,66 @@
-# payments
+## Foundry
```
Revert this change?

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,418 @@
+        payer.lockupRate = payer.lockupRate - oldRate + rate;
+
+        // Init hasLockupStarted flag and lockupStart for the first rail with non-zero rate
+        if (rate > 0 && !payer.hasLockupStarted) {
```
Take a look at the "Lockup can be settled..." part of: https://www.notion.so/filecoindev/Payments-Design-Iteration-17ddc41950c180468806c7e0097c2e14?pvs=4#185dc41950c1805da8d4ce567e343307

`lockupStart` is probably misnamed. Really, it means "the last time we settled lockup". I.e., it's when the lockup rate starts applying.

My point is, I don't think we need `hasLockupStarted`. We always adjust `lockupStart`.

Stebalien replied:
Ah, I see you've implemented `accumulateLockupRate`. You don't need this code here, you just need to call `accumulateLockupRate` before changing the rate (e.g., at the very top of this function). The first time this is called, that will accumulate nothing because the rate will be 0, but it will initialize `lockupStart`.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+
+        Account storage toAccount = accounts[token][to];
+        require(toAccount.ownerAddress != address(0), "to account does not exist");
+        require(toAccount.funds > 0, "to account has no funds");
```
They don't need funds, do they?

aarshkshah1992 replied:
@Stebalien I mean not strictly speaking. We can get rid of this yeah.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+
+        Account storage fromAccount = accounts[token][from];
+        require(fromAccount.ownerAddress != address(0), "from account does not exist");
+        require(fromAccount.funds > 0, "from account has no funds");
```
Technically, the sender doesn't need any funds either.

aarshkshah1992 replied:
@Stebalien Technically no but I'd like to keep this so that rails are only created with clients who have interacted with the contract before. Just think it's better that way.

Stebalien replied:
Well, this goes back to who can create a rail. IMO, the answer is: the client.

It would be nice if a client can go through the entire signup flow without having to deposit any funds until they actually start using the service.

aarshkshah1992 replied:
We agree that neither `from` or `to` needs funds.

aarshkshah1992 replied:
Done.

@Stebalien -> Not only do they not need funds now, but we've fixed this so that both `from` and `to` accounts will be created if they don't already exist.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        require(fromAccount.ownerAddress != address(0), "from account does not exist");
+        require(fromAccount.funds > 0, "from account has no funds");
+
+        if (approval.arbitrer != address(0)) {
+            require(arbiter == approval.arbitrer, "arbiter mismatch");
+        }
```
TBD: need to discuss approved operators and who, precisely, is allowed to call `createRail`.

aarshkshah1992 replied:
We agreed that only operators can create rails.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+
+        uint256 railId = _nextRailId++;
+
+        Rail storage rail = rails[railId];
```
OT: Ethereum's "everything exists, it just defaults to zero" gets me every time. It's kind of cool but also really disconcerting.

aarshkshah1992 replied:
@Stebalien What's "OT" ?

Stebalien commented on .gitignore:
```diff
@@ -0,0 +1,14 @@
+out/
+
+# Ignores development broadcast logs
+!/broadcast
+/broadcast/*/31337/
+/broadcast/**/dry-run/
```
Why not just ignore all of /broadcast?

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+    struct ArbitrationResult {
+        bool approved;
+        uint256 modifiedAmount;
+        uint256 settledUntil;
```
We should consider adding an arbitrary "note" field. Could be useful to explain why the full amount isn't approved.

aarshkshah1992 replied:
Added this and also now returning this to the caller of `settleRail`

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+
+interface IArbiter {
+    struct ArbitrationResult {
+        bool approved;
```
Do we need `approved`? IMO, we should just revert if we're not approving anything, but I'm not really a solidity expert.

aarshkshah1992 replied:
@Stebalien This is for the Arbitrer contract to gracefully let us know that it is/is not rejecting the payment. The Payment contract does revert the settlement if it sees a disapproval from the arbitrer.

aarshkshah1992 replied:
The other option is to not gracefully handle any reverts from the `Arbitrer` contract at all and simply revert right away (in Solidity, when contract X calls contract Y and contract Y reverts -> the call to X reverts right away unless you handle it as above).

The way we've implemented it here -> any revert from `Arbitrer` will bubble up right away and the settlement will revert but atleast for payment refusals -> we get to handle it gracefully and do more stuff if we want to before we revert.

aarshkshah1992 replied:
The Arbitrer SHOULD be able to revert if it wants to.

aarshkshah1992 replied:
Done. Removed this.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        address ownerAddress;
+        uint256 funds;
+        uint256 lockupCurrent;
+        uint256 lockupRate;
```
Let's comment on the units here (and all the other fields). E.g. lockup rate is attotoken/epoch.

aarshkshah1992 replied:
@Stebalien Yeah plan is to raise a subsequent PR for comprehensive docs for both state and functions once we've locked in a meaningful draft.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        uint256 paymentRate;
+        uint256 lockupPeriod;
+        uint256 lockupFixed;
+        uint256 lastSettledAt;
```
Hm. Now that I see this, I wonder if we should call this "settledUpTo" instead of "lastSettledAt".

aarshkshah1992 replied:
Also renamed the corresponding `ArbitrationResult` field to `settledUpTo`

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+    }
+
+    struct OperatorApproval {
+        bool isApproved;
```
Do we need this field?

aarshkshah1992 replied:
@Stebalien Yeah it's the easiest way to track the fact that an operators has been disapproved. Solidity HashMap/Array deletes aren't clean/gas cheap to do.

Stebalien replied:
Also, if _operators_ create rails, it makes sense to:

1. Approve the operator with no allowance.
2. Let the operator create rails.
3. Then give them allowance.

(IMO)

aarshkshah1992 replied:
Have changed `createRail` to automagically approve an operator with zero allowance if it does not already exist.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+
+    struct OperatorApproval {
+        bool isApproved;
+        address arbitrer;
```
Do we need this field either? Isn't it implicit?

aarshkshah1992 replied:
@Stebalien My idea was to have the client have a say in who gets to be an arbitrer.

aarshkshah1992 replied:
We agree that we shouldn't have this here as the arbitrer might be different for each rail.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        return (totalSettled, finalEpoch);
+    }
+
+    /// @notice Returns the minimum of two numbers.
+    function min(uint256 a, uint256 b) internal pure returns (uint256) {
+        return a < b ? a : b;
+    }
```
Make this a free function?

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+            totalSettled += segmentAmount;
+            currentEpoch = segmentSettledEpoch;
+            rateQueue.dequeue();
+            }
```
```suggestion
        }
```

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        require(settledUntil > segmentStart, "no progress made in settlement");
+
+        return (settledAmount, settledUntil);
+    }
+    function _settleWithRateChanges(
```
```suggestion
    }

    function _settleWithRateChanges(
```

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        uint256 oldLockup = rail.lockupFixed + (rail.paymentRate * rail.lockupPeriod);
+        uint256 newLockup = lockupFixed + (rail.paymentRate * period);
+
+        // assert that operator allownace is respected
```
```suggestion
        // assert that operator allowance is respected
```

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+                segmentStart,
+                segmentEnd
+            );
+            require(arbResult.approved, "arbitrer refused payment");
```
We need to validate that:

1. `arbResult.settledUntil <= settledUntil`.
2. `arbResult.modifiedAmount <= rate * (arbResult.settledUntil - segmentStart)`.

aarshkshah1992 replied:
For 1, it should be `arbResult.settledUntil <=segmentEnd`, right ? Yeah making this change.

aarshkshah1992 replied:
Done.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+
+        settleAccountLockup(payer);
+
+        uint256 settlementTargetEpoch = min(untilEpoch, payer.lockupLastSettledAt + rail.lockupPeriod);
```
Need to check that `untilEpoch` is at or before the current epoch.

aarshkshah1992 replied:
Done.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+            uint256 segmentRate = currentRate;
+            if (!rateQueue.isEmpty()) {
+                RateChangeQueue.RateChange memory nextChange = rateQueue.peek();
+                if (nextChange.untilEpoch > currentEpoch && nextChange.untilEpoch <= targetEpoch) {
+                    nextBoundary = nextChange.untilEpoch;
+                    segmentRate = nextChange.rate;
+                }
```
I've spent a lot of time staring at this code and I'm still not sure if it's right. Let's talk through it tomorrow.

aarshkshah1992 replied:
I agree. It is involved.

aarshkshah1992 replied:
Let me try to simplify it more today before we sync up.

aarshkshah1992 replied:
Have refactored this entire flow to make it easy to reason about.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+    uint256 private _nextRailId;
+
+    // token => owner => Account
+    mapping(address => mapping(address => Account)) public accounts;
```
@Stebalien 

Can't we implement those getters as view functions that can be invoked off-chain on RPC providers ? We can add some state to the contract that makes those view functions slightly less compute intensive but IMO making it a view function that can be invoked off-chain means we can offload the worrying about efficiency to client <> RPC provider relationships without having to add too much state to the contract.

Stebalien replied:
Ok, we can solve (1) by:

1. Adding a `mapping(address => address[])` of account owners to tokens.
2. Optionally: changing `token => owner => Account` to `owner => token => Account`. This isn't strictly necessary, but it mirrors the mapping in (1).

As discussed, we can drop (2) and let the user handle indirection through smart contracts.

Stebalien replied:
Actually, maybe we want a single `owner => Account` mapping where `Account` contains:

1. The client's operator approvals (map of `operator => Approval`).
2. An array of approved client operators.
3. An array of client rail IDs.
4. etc...

Stebalien replied:
Resolution: we're still exploring the API, let's build that first then figure out what these data structures should look at.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+    function approveOperator(
+        address token,
+        address operator,
+        address arbiter,
```
@Stebalien You mean the client creates the rail OR that a client has to approve a rail that has been created by an operator ? Both have different UX implications.

aarshkshah1992 replied:
See my comment at https://github.com/FilOzone/payments/pull/1/files#r1961082248.

aarshkshah1992 replied:
An operator can have many different arbitrers for different rails.

aarshkshah1992 replied:
So remove this.

aarshkshah1992 replied:
The answer here is to trust the operator.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,572 @@
+        address to,
+        address operator,
+        address arbiter
+    ) external returns (uint256) {
```
@Stebalien It does make the UX harder for clients and operators. It's more intuitive for an operator to create a rail as they know who the provider of the service is etc (without always leaking that to the client).

If we get clients to create a rail -> we need to add a step where the operator has to approve the rail and the operator needs to wait/watch for rail creation before they can start providing services. It really does complicate the entire flow.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        bool approved;
+        uint256 modifiedAmount;
+        uint256 settleUpto;
+        string note;
```
Can leave a note if successful too. For better UX, it's a catch all field for arbitrer to send us info.

aarshkshah1992 replied:
This is done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        
+    }
+
+    function updateRailArbiter(uint256 railId, address newArbiter) external validateRailActive(railId) onlyRailOperator(railId) {
```
Do we really need this ?

aarshkshah1992 replied:
Removed  this. Will add it back if needed.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        address token,
+        address from,
+        address to,
+        address operator,
```
we don't need this as the `msg.sender` will ALWAYS be the operator.

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+    }
+
+    // TODO: Revisit
+    function terminateOperator(address operator) external  {
```
who calls this ?

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+// SPDX-License-Identifier: UNLICENSED
```
Write down all the `getter` APIs.

aarshkshah1992 replied:
Try to make it as easy as possible to access basic state without an Indexer.

aarshkshah1992 replied:
Strike a balance between on-chain state and off-chain compute than an RPC provider needs to do to serve Getter requests.

aarshkshah1992 replied:
Storing the keys of a Hasmap in an a set if you need iteration. Do it only when needed.

aarshkshah1992 replied:
Added license and made the changes.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+    }
+
+    // TODO: implement
+    function terminateRail() public {
```
An operator can always do this by setting everything to 0.

aarshkshah1992 replied:
If a client wants to terminate a rail -> they go the operator and get them to modify the rates and lockups for the rail to 0.

aarshkshah1992 replied:
Removed this function.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+    }
+
+    struct Rail {
+        bool isActive;
```
We need this to track rail termination !

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+
+        Rail storage paymentRail = rails[railId];
+        Account storage payer = accounts[paymentRail.token][paymentRail.from];
+        Account storage payee = accounts[paymentRail.token][paymentRail.to];
```
Does the solidity compiler optimise this so we only lookup the map once ?

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        settleAccountLockup(payer);
+
+        // Determine the maximum allowed epoch we can settle to based on the payer's lockupPeriod.
+        uint256 maxSettlementEpoch = min(untilEpoch, payer.lockupLastSettledAt + paymentRail.lockupPeriod);
```
can read ` payer.lockupLastSettledAt` from return value.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        uint256 currentPaymentRate = paymentRail.paymentRate;
+
+        // Retrieve any queued rate changes for this rail.
+        RateChangeQueue.Queue storage rateQueue = railRateChangeQueues[railId];
```
We can put the `railRateChangeQueues ` in the `Rail` itself.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        uint256 railId,
+        Rail storage rail,
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 initialRate,
```
this is the current rail rate. fix naming.

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        uint256 activeRate = initialRate;
+        note = "";
+
+        while (currentEpoch <= targetEpoch) {
```
currentEpoch < targetEpoch is what we need here. if it is equal, we are done and we should exit the loop.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+                RateChangeQueue.RateChange memory upcomingChange = rateQueue.peek();
+                bool isWithinRange = (
+                    upcomingChange.untilEpoch >= currentEpoch &&
+                    upcomingChange.untilEpoch <= targetEpoch
```
This second condition won't be true if I'm trying to settle up to some historical epoch in the past.

aarshkshah1992 replied:
We need to remove this assert.

aarshkshah1992 replied:
Done.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+
+            // Otherwise, we have fully settled up to nextBoundary.
+            currentEpoch = segmentEndEpoch;
+            note = arbNote;
```
It would be kind of nice to somehow join these notes together when non-empty, but that isn't critical. We mostly care about the last note anyways.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+            // Dequeue the rate change we just passed (if it was indeed used this round).
+            if (!rateQueue.isEmpty()) {
+                // Peek again to confirm it's the one we used.
+                RateChangeQueue.RateChange memory frontChange = rateQueue.peek();
```
How would this not be the case?

aarshkshah1992 replied:
if the queue is non-empty AND next element has expired -> until epoch is in the past or < targetEpoch -> pop.

aarshkshah1992 replied:
actually -> we can JUST dequeue if we reach here.

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+                    upcomingChange.untilEpoch <= targetEpoch
+                );
+                require(isWithinRange, "rate queue is in an invalid state");
+                nextBoundary = upcomingChange.untilEpoch;
```
min(next, target)

aarshkshah1992 replied:
nextBoundary = min(nextBoundary, target)

aarshkshah1992 replied:
Done.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+                railId,
+                rail,
+                currentEpoch,
+                nextBoundary,
```
pass `min(nextBoundary, targetEpoch)`

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        return (settledAmount, settledUntil, note);
+    }
+
+    function _settleWithRateChanges(
```
mark the function as `nonReEntrant`.

aarshkshah1992 replied:
Infact all functions that modify rail state should have a per-rail lock.

aarshkshah1992 replied:
Audit and REALLY think through it.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+     * @return settledUntil The final epoch that effectively got settled.
+     * @return note An optional note returned by the arbitrator.
+     */
+    function _settleSegment(
```
rentrancy

aarshkshah1992 replied:
per-rail settlement lock

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+    }
+
+    // railId => RateChangeQueue
+    mapping(uint256 => RateChangeQueue.Queue) public railRateChangeQueues;
```
move this to the Rail struct.

aarshkshah1992 replied:
Done.

Stebalien commented:
The re-entency issue:

We discussed a few approaches:

1. Lock everything preventing both read and write access to the entire contract while settling a rail. This would restrict access to important information, e.g., information about the rail.
2. Lock the rail only. Unfortunately, if we lock read access, we'd still be restricting information about the rail. Furthermore, if we don't adjust payee/payer funds while settling and still allow, e.g., the arbiter to _see_ those fields, the arbiter won't see the "full" effects of the last section of payment it arbitrated.

We settled on:

1. Locking modifications (preventing settling, changing lockup, etc.) to the rail while settling.
2. Updating the rail, the payee, and the payer after _each_ call to the arbiter such that the arbiter always sees the payment contract in the correct state. Effectively, we move most of the rail settlement logic into the "settle segment" function.

Thinking on this more, there _are_ a few more solutions (which I don't like but I figured I'd mention for completeness).

1. We _could_ not lock anything, record important state before calling the arbiter, call the arbiter, then revert if "something has changed". This "works" but is strictly more brittle and doesn't provide any benefit.
2. We _could_ allow payment/lockup changes as long as we're really careful as those changes can only affect future time periods. But I don't think it's worth it and re-entrant modifications are likely always bugs.

Stebalien commented:
We also discussed how this relates to withdrawal. IMO, the difference between this and withdrawal is that all the "withdraw" call can do is revert. It can't cause us to change our minds and withdraw _less_ than we initially intended so there are no funky intermediate states.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,612 @@
+        uint256 activeRate = initialRate;
+        note = "";
+
+        while (currentEpoch <= targetEpoch) {
```
Done.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+            note = result.note;
+
+            // If arbiter made no progress, return early without updating any state
+            if (settledUntilEpoch <= epochStart) {
+                return (0, note);
+            }
+
+            // Ensure arbiter doesn't allow more payment than the maximum possible
+            // for the epochs they're confirming
+            uint256 maxAllowedAmount = rate * (settledUntilEpoch - epochStart);
+            require(
+                result.modifiedAmount <= maxAllowedAmount,
+                "arbiter modified amount exceeds maximum for settled duration"
+            );
```
I'd:

1. Require that `settledUntilEpoch >= epochStart`.
2. Require that the amount settled is within expectations given the `settledUntilEpoch`.
3. Finally, short-circuit if no funds were actually settled.  Or, TBH, I'd just go to the end of the function anyways and not short-circuit.

Specifically, I don't want to return `(0, note)` if the arbiter returns nonsensical results (e.g., settle 1M FIL up till some epoch in the past).

aarshkshah1992 replied:
This is done.

aarshkshah1992 replied:
TODO for @aarshkshah1992 : Audit the code for off by ones to enforce that the `untilEpoch` in the rate-change queue is not inclusive of the end epoch and that `rail.settledUpto` is inclusive of the settledEpoch.

aarshkshah1992 replied:
Will do this in the follow-up PR.

Stebalien commented:
This is looking really good. I'd say punt operator termination stuff to a new PR, finish up the last nits here, then we can merge it.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+        _;
+    }
+
+    modifier noRailModificationInProgress(uint256 railId) {
+        require(
+            !_railModificationLocks[railId],
+            "Modification already in progress"
+        );
+        _railModificationLocks[railId] = true;
+        _;
+        _railModificationLocks[railId] = false;
+    }
```
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+            require(
+                settledUntilEpoch == block.number,
+                "Failed to settle rail completely"
+            );
```
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+        }
+    }
+
+    function deposit(address token, address to, uint256 amount) external {
```
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+    )
+        external
+        validateRailActive(railId)
+        validateRailAccountsExist(railId)
```
I don't think we need the ` validateRailAccountsExist` modifier at all now as all accounts always exist implicitly with zero funds (we've now removed the `ownerAddress` field from Account).

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+    )
+        external
+        validateRailActive(railId)
+        validateRailAccountsExist(railId)
```
I don't think we need the  validateRailAccountsExist modifier at all now as all accounts always exist implicitly with zero funds (we've now removed the ownerAddress field from Account).

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+        // Create account if it doesn't exist
+        Account storage account = accounts[token][to];
+        if (account.ownerAddress == address(0)) {
+            account.ownerAddress = to;
```
Yeah, you are right. Removing this field and also removing the `onlyAccount` modifier simplifies things a bit and we can always infer the account owner as being the msg sender for any methods that need the sender to prove account ownership.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,694 @@
+        settleAccountLockup(account);
+    }
+
+    function withdraw(address token, uint256 amount) external nonReentrant {
```
@Stebalien 

I have also implemented a `function withdrawTo(
        address token,
        address to,
        uint256 amount
    )`
    
    function that allows users to withdraw funds to any account of their choosing.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+            approval.isApproved = true;
+            approval.rateAllowance = 0;
+            approval.lockupAllowance = 0;
+        }
```
Done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+            rail.from
+        ][rail.operator];
+
+        require(approval.isApproved, "operator not approved");
```
@Stebalien Let's talk about this in our 1:1 today.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+                "not able to settle rail upto current epoch"
+            );
+        } else {
+            rail.rateChangeQueue.enqueue(rail.paymentRate, block.number);
```
@Stebalien Good catch ! How do we want to implement this ?

If a user calls `modifyRailPayment` multiple times in a single epoch -> should we only enque the previous (before this epoch) `rail.paymentRate ` ONLY once and ensure that `rail.paymentRate ` reflects the last rate update ?

Basically, the rate changes within a single epoch do not end up on the queue.

aarshkshah1992 replied:
@Stebalien I also realised after looking at this again is that we're basically saying that the new rate will be applicable AFTER the current block and rail will be settled upto and including this epoch using the old rate. 

Just wanted to confirm if that's the expected behaviour. IMO, it makes sense.

Stebalien replied:
> Basically, the rate changes within a single epoch do not end up on the queue.

Exactly.

> Just wanted to confirm if that's the expected behaviour. IMO, it makes sense.

IMO we could do it either way but... I'd keep it the current way as it's simpler.

aarshkshah1992 replied:
This is done.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -0,0 +1,793 @@
+
+    function withdraw(
+        address token,
+        uint256 amount
```
@Stebalien Implemented this.

aarshkshah1992 commented:
@Stebalien 

I've addressed all your comments.

Only one comment remains unresolved at https://github.com/FilOzone/payments/pull/1#discussion_r1976817452. Let's go over it in our sync today and we can then merge it.

I am now starting work on raising a new PR to implement operator termination and all the other design changes we've added to  https://www.notion.so/filecoindev/Payments-Design-Iteration-17ddc41950c180468806c7e0097c2e14?pvs=4 since our first draft.

aarshkshah1992 commented:
@Stebalien Fixed the CI and removed the `configure` script.

Stebalien commented on src/RateChangeQueue.sol:
```diff
@@ -0,0 +1,50 @@
+        uint256 tail;
+    }
+
+    function enqueue(Queue storage queue, uint256 rate, uint256 untilEpoch) internal {
```
TODO: drop enqueued rate-changes if `queue.changes[queue.tail].untilEpoch == untilEpoch`.

aarshkshah1992 replied:
@Stebalien I want to keep the queue implementation simpler. I've made this change at the call site in `modifyRailPayment`.

Stebalien commented on src/Payments.sol:
```diff
@@ -0,0 +1,700 @@
+
+        // --- Settlement Prior to Rate Change ---
+        // If there is no arbiter, settle the rail immediately.
+        if (rail.arbiter == address(0)) {
+            (, uint256 settledUpto, ) = settleRail(railId, block.number);
+            require(
+                settledUpto == block.number,
+                "not able to settle rail upto current epoch"
+            );
+        } else {
+            rail.rateChangeQueue.enqueue(rail.paymentRate, block.number);
```
We should probably skip all of this if the rate hasn't changed at all (if we're _just_ making a one-time payment. But it should be a no-op otherwise so I'm happy doing this in a new PR.

aarshkshah1992 replied:
Doing this in the next PR.

Stebalien reviewed (approved):
LGTM. There are some optimizations we can make, but they aren't required for correctness.

Stebalien commented:
I'll let you press the merge button when you wake up.

aarshkshah1992 commented on src/Payments.sol:
```diff
@@ -326,7 +326,14 @@ contract Payments is
+            // handle multiple rate changes by the operator in a single epoch
+            // by queueuing the previous rate only once
+            if (
+                rail.rateChangeQueue.isEmpty() ||
```
@Stebalien FYI
